Add commands to install, list, search, update and uninstall plugins

uberhacker opened the pull request:
- Refactored plugin commands into separate class files as per @greg-1-anderson's suggestion

uberhacker committed:
Add PluginCommand to install, list, uninstall and update plugins

uberhacker committed:
Merge branch 'master' into master

Commit:
Update to fix code sniffer issues

uberhacker committed:
Add ability to install plugins via 'composer create-project'

Commit:
Refactored plugin commands into separate class files

Commit:
Update to fix TODO notes

  - Change messages to be more consistent
  - Clean up docblocks

uberhacker commented:
Testing help anyone?  @TeslaDethray, @ronan, @greg-1-anderson?

greg-1-anderson commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,334 @@
+     */
+    protected function getLatestVersion($plugin)
+    {
+        exec("cd \"$plugin\" && git fetch --all && git tag -l | sort -r | head -1", $tag);
```
This should only consider tags that are (a) semver and (b) the same major version as Terminus' major version.

uberhacker replied:
Agreed.  I'll work on limiting the results of the tag listing.

greg-1-anderson commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,334 @@
+                $project = $composer_info['name'];
+                $packagist_url = "https://packagist.org/packages/{$project}";
+                if ($this->isValidUrl($packagist_url)) {
+                    exec("rm -rf \"{$plugin_dir}\" && composer create-project --prefer-source --keep-vcs -n -d {$plugins_dir} {$project}:~1", $messages);
```
Since you are installing with `--prefer-source` and `--keep-vcs`, you could update using the same technique used to do a "git" update above.

uberhacker replied:
That would be true if the initial install was via `--prefer-source --keep-vcs`.  Don't be confused by the `composer` method.  It basically means a `.git` directory could not be found.

greg-1-anderson replied:
Oh, okay, got it. So if you use the install command to add the plugin, and it runs `composer create-project...`, then the update command will follow the `git` branch here. That makes sense. It might be beneficial to continue to support multiple update methods here, even if only one installation method is supported, just to make things easy for folks who may have installed without the tool.

greg-1-anderson replied:
Doing the `rm -rf` prior to the `composer create-project` could be pretty dangerous in some situations, e.g. if someone was writing their own plugin that had no git repo / packagist project, and accidentally named it in an `update` command. It would be very sad if they ended up in this code path and deleted their work. Maybe moving the plugin out of the way to `~/.composer/backups` would be safer.

uberhacker replied:
We should recommend developers always use git when creating their own plugin. :)  However, you have a good point.

greg-1-anderson commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,334 @@
+                    if (!empty($composer_info)) {
+                        $project = $composer_info['name'];
+                    }
+                    $archive_url = "https://github.com/{$project}/archive/1.x.tar.gz";
```
There is no guarentee that the project came from GitHub, or that it used the 1.x branch.

uberhacker replied:
Right, but there is no way of really knowing for sure since we don't store that anywhere after an install.  If it can't be found via github.com, then it will just send a message that the project doesn't exist.

greg-1-anderson commented:
There's a lot of useful code here. However, I think only one install technique should be implemented; it's too much (or at least unnecessary) to maintain all three. My preference is still `composer create-project`, with `git` being a close second. I don't think we should install via `curl | tar`, because that always installs the latest dev version of the plugin. If we did that, then it wouldn't be possible for plugin developers to have releases unless they did something unusual, such as keeping their `dev` code on a separate branch -- or something inconvenient, like changing the install URL in the README on every release.

uberhacker commented:
I like the idea of using a "standard" method of installation.  We could install via `composer create-project --prefer-source --keep-vcs` and update via `git checkout`.  The nice thing about composer is it allows you to search for plugins at packagist.org.

Commit:
Merge branch 'master' of github.com:pantheon-systems/terminus

Commit:
Fix to get the major version based on the installed Terminus release

  - Backup plugin projects that do not have a .git directory during updates

greg-1-anderson commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,356 @@
+                    $backup_directory = "~/.composer/backups/{$project}/{$datetime}";
+                    exec("mkdir -p {$backup_directory} && tar czvf {$backup_directory}/backup.tar.gz \"{$plugin_dir}\"", $backup_messages);
+                    // Create a new project via Composer.
+                    exec("rm -rf \"{$plugin_dir}\" && composer create-project --prefer-source --keep-vcs -n -d {$plugins_dir} {$project}:~1", $install_messages);
```
The `:~1` here should also be composed from Terminus major version

uberhacker replied:
Good call.  Missed that.  I'll try and push up another commit soon that fixes this and a few other issues.

Commit:
Add the 'Terminus major version' variable in all relevant locations

  - Remove archive installs from updates
  - Change project backup location

uberhacker committed:
Merge branch 'master' of github.com:pantheon-systems/terminus

uberhacker committed:
Add requirements check

  - Remove obsolete or unused methods
  - Provide composer only install option

greg-1-anderson commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,318 @@
+         $result = fgets($file, 255);
+         $valid = $windows ? !preg_match('#Could not find files#', $result) : !empty($result);
+         if (!$valid) {
+            die("Please install the git command to enable plugin management.\n");
```
Never use `die()` -- this method is not testable. Instead, typically you should throw an exception (e.g. TerminusException). However, you should not fail like this in a command constructor, as failures here will prevent the operation of any command.

uberhacker committed:
Merge branch 'master' of github.com:pantheon-systems/terminus

uberhacker committed:
Refactor requirements check

greg-1-anderson commented on src/Commands/Plugin/UpdateCommand.php:
```diff
@@ -23,6 +25,16 @@ class UpdateCommand extends PluginBaseCommand
      */
     public function update(array $plugins)
     {
+        // Check for minimum plugin command requirements.
```
Instead of duplicating the messages et. al. for commandExists in each command, you could instead further factor this into a checkRequirements method that does the command checking and exception throwing. Also, I think there's little value in allowing someone to install but not update w/out installing git. You could just check for git and composer for all commands. Keep it simple.

uberhacker committed:
Refactor requirements check

  - Reformat search results output

Commit:
Add empty project check with usage text to install command

Commit:
Fix coding standards issue with SearchCommand.php

greg-1-anderson committed:
Touch-up plugin:search command.

Commit:
Fix merge conflict in SearchCommand.php

Commit:
Rename $approved to $approvedProjects in SearchCommand.php line 76

Commit:
Add \@param to docblock of checkStatus method in SearchCommand.php

Commit:
Remove unreferenced use line in SearchCommand.php

Commit:
Bypass Symfony exception and display something more understandable when there is no keyword supplied with the plugin:search command

Commit:
Add stability option to InstallCommand.php

Commit:
Fix @option stability in InstallCommand.php

greg-1-anderson committed:
Register pancakes as an 'Approved' plugin. Rename 'Unknown' to 'Unofficial'.

uberhacker committed:
Merge branch 'master' of https://github.com/uberhacker/terminus

ronan commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,335 @@
+     * @param string $plugin Plugin name
+     * @return string Plugin directory
+     */
+    protected function getPluginDir($plugin = '')
```
All this shouldn't be needed anymore. Just make this command 'configaware' and call: `$this->getConfig()->get('plugins_dir');`

uberhacker replied:
@ronan: I appreciate the feedback.  The rub here is this function returns the entire path of the plugin, not just the `~/.terminus/plugins` directory.  I suppose we could still use this function as a wrapper around `$this->getConfig()->get('plugins_dir');` instead of including all the extra logic that determines the base directory.

uberhacker committed:
Merge branch 'master' of github.com:pantheon-systems/terminus

uberhacker committed:
Merge branch 'master' of github.com:pantheon-systems/terminus

uberhacker committed:
Update for better Windows support

  - Move updatePlugin method from PluginBaseCommand.php to
  UpdateCommand.php
  - Remove getSlash method in PluginBaseCommand.php

uberhacker committed:
Merge branch 'master' into master

uberhacker committed:
Remove PluginCommand.php

uberhacker committed:
Merge branch 'master' of github.com:uberhacker/terminus

uberhacker committed:
Add DIRECTORY_SEPARATOR to the getPluginProjects method in PluginBaseCommand.php

uberhacker committed:
Add smarter sorting of version tags, search and list results

uberhacker commented:
@TeslaDethray: Do we have a replacement solution on the roadmap?

uberhacker commented:
@TeslaDethray: Accidentally closed? LoL  I was going to attempt to write tests at some point but since you volunteered, I was waiting to see what you came up with in the meantime.

greg-1-anderson commented:
Others may still wish to push this forward.

uberhacker commented:
Thanks @greg-1-anderson.  Any others besides you?  :-P

uberhacker commented:
I hear crickets...

uberhacker committed:
Update to the latest stable release if a previous non-stable semantic version is available

uberhacker committed:
Add semverCompare method to correctly compare semantic versions

uberhacker committed:
Replace semverCompare method with Composer\Semver\Comparator

iamEAP commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,241 @@
+    {
+        // Get the Terminus major version.
+        $terminus_major_version = $this->getTerminusMajorVersion();
+        exec("cd \"$plugin\" && git fetch --all && git tag -l | grep ^[v$terminus_major_version] | sort -Vr | head -1", $tag);
```
Writing tests now and caught a problem here. OS X's default `sort` command does not support `-V`.  Work-around is to apply the sorting on the `git tag` command. As of Git 2.0 (released in 2014), should be able to `git tag -l --sort=version:refname`

SaraMcCutcheon replied:
I've allowed this to go for so long such that there is another contender for these commands, submitted the same day as this update. I'll need to take another look at the two side by side.
https://github.com/pantheon-systems/terminus/pull/1759

uberhacker replied:
Hello @TeslaDethray!  Feel free to use whatever method seems best.  I expected that we would need to change some things around before it was accepted.  If it helps any, I can give you access to edit this branch.  I gave @greg-1-anderson permission to make changes awhile ago.  Welcome back! :)

greg-1-anderson replied:
It looks like #1759 is a continuation of this PR.

iamEAP replied:
Yep.  Would recommend consolidating, either by cherry-picking or merging.

SaraMcCutcheon commented on src/Commands/Plugin/PluginBaseCommand.php:
```diff
@@ -0,0 +1,241 @@
+        $windows = (php_uname('s') == 'Windows NT');
+        $test_command = $windows ? 'where' : 'command -v';
+        $file = popen("$test_command $command", 'r');
+        $result = fgets($file, 255);
```
Would using Finder components help alleviate the need to check for system type at runtime? (Also helps w/ the testing)

uberhacker replied:
Sure, if you want to add some code snippet suggestion as a comment, I can add it or you can if you want permission to edit this branch.

uberhacker committed:
Fix MacOS related bug in 'terminus plugin:search' as reported by @cleverington

SaraMcCutcheon commented:
Closing due to age, supercession, and a lack of tests.
